File: backend/src/main/java/com/kdnth/campaignkeep/spell/Spell.java
package com.kdnth.campaignkeep.spell;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "spells")
@Getter
@Setter
@NoArgsConstructor
public class Spell {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "api_index", unique = true, length = 100)
    private String apiIndex;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(nullable = false)
    private Short level;

    @Column(name = "casting_time", length = 50)
    private String castingTime;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "range_type", nullable = false)
    private SpellRangeType rangeType;

    @Column(name = "range_feet")
    private Short rangeFeet;

    @Column(length = 10)
    private String components;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(nullable = false)
    private MagicSchool school;

    @Column(nullable = false)
    private Boolean concentration = false;

    @Column(name = "has_verbal", nullable = false)
    private Boolean hasVerbal = false;

    @Column(name = "has_somatic", nullable = false)
    private Boolean hasSomatic = false;

    @Column(name = "has_material", nullable = false)
    private Boolean hasMaterial = false;

    @Column(name = "material_desc", columnDefinition = "TEXT")
    private String materialDesc;

    @Column(nullable = false)
    private Boolean ritual = false;

    @Column(length = 100)
    private String duration;

    @Column(name = "higher_level", columnDefinition = "TEXT")
    private String higherLevel;
}

File: frontend/src/stores/spell.ts
import apiClient from '@/api/axios'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export type MagicSchool =
  | 'abjuration'
  | 'conjuration'
  | 'divination'
  | 'enchantment'
  | 'evocation'
  | 'illusion'
  | 'necromancy'
  | 'transmutation'

export type SpellRangeType = 'self' | 'touch' | 'ranged' | 'special' | 'sight' | 'unlimited'

export interface SpellSummaryResponse {
  id: number
  apiIndex: string
  name: string
  level: number
  school: MagicSchool
  castingTime: string
  concentration: boolean
  ritual: boolean
  hasVerbal: boolean
  hasSomatic: boolean
  hasMaterial: boolean
}

export interface SpellDetailResponse extends SpellSummaryResponse {
  duration: string | null
  rangeType: SpellRangeType
  rangeFeet: number | null
  rangeDisplay: string
  materialDesc: string | null
  description: string | null
  higherLevel: string | null
}

export interface SpellListParams {
  search?: string
  level?: number
  school?: MagicSchool
  classIndex?: string
  maxLevel?: number
}

export const useSpellStore = defineStore('spell', () => {
  const spells = ref<SpellSummaryResponse[]>([])
  const characterSpells = ref<SpellDetailResponse[]>([])
  const activeSpell = ref<SpellDetailResponse | null>(null)
  const isSyncing = ref(false)

  function clearSpells() {
    spells.value = []
    characterSpells.value = []
    activeSpell.value = null
  }

  async function fetchSpells(params: SpellListParams = {}) {
    isSyncing.value = true
    try {
      const response = await apiClient.get<SpellSummaryResponse[]>('/spells', { params })
      spells.value = response.data
      return response.data
    } finally {
      isSyncing.value = false
    }
  }

  async function fetchSpell(id: number) {
    const response = await apiClient.get<SpellDetailResponse>(`/spells/${id}`)
    activeSpell.value = response.data
    return response.data
  }

  async function fetchCharacterSpells(characterId: number) {
    const response = await apiClient.get<SpellDetailResponse[]>(`/characters/${characterId}/spells`)
    characterSpells.value = response.data
    return response.data
  }

  async function addSpellToCharacter(characterId: number, spellId: number) {
    const response = await apiClient.post<SpellDetailResponse>(`/characters/${characterId}/spells`, {
      spellId,
    })
    if (!characterSpells.value.some((spell) => spell.id === spellId)) {
      characterSpells.value.push(response.data)
    }
    return response.data
  }

  async function removeSpellFromCharacter(characterId: number, spellId: number) {
    await apiClient.delete(`/characters/${characterId}/spells/${spellId}`)
    characterSpells.value = characterSpells.value.filter((spell) => spell.id !== spellId)
  }

  return {
    spells,
    characterSpells,
    activeSpell,
    isSyncing,
    clearSpells,
    fetchSpells,
    fetchSpell,
    fetchCharacterSpells,
    addSpellToCharacter,
    removeSpellFromCharacter,
  }
})
